### PR TITLE
Banner Picker

### PR DESCRIPTION
**Pull request by GHaberis**
## What does this change?

Currently there are 6 banners competing to be shown in frontend.

`javascripts/projects/common/modules/ui/first-pv-consent-banner.js`
`javascripts/projects/common/modules/ui/cookiesBanner.js`
`javascripts/projects/common/modules/ui/smartAppBanner.js`
`javascripts/projects/common/modules/commercial/membership-engagement-banner.js`
`javascripts/projects/common/modules/navigation/membership.js`
`javascripts/projects/common/modules/identity/global/opt-in-engagement-banner.js`

Each of these modules contains conditional checks which have to be met before they create a new instance of `Message.js` and call `Message.show()`.

Currently If the conditions for multiple banners are satisfied  then multiple instances of `Message` are setup (1 for each banner), and `Message.show` is called for each instance. 

A check within the `Message.show` checks to see whether a message has already been added to the DOM and prevents another banner being added if so.

**Why is this a problem?**

- The banner which is shown is the banner which calls `Message.show` first.
- There's no way to prioritise a high-priority banner which has async checks in it's conditions. This means even if you initialise the high-priority banner first it won't get shown if another banner has passed it's own conditions and called `Message.show` first.

**How does this PR fix this?**

- I've created a module `bannerPicker.js`. The `init` method takes a single argument - a list of banners ordered by priority.
- bannerPicker is initialised from `common/enhanced` 
-`bannerPicker` will execute the `canShow()` check for each banner, and `show()` for the highest priority banner whose checks have passed.
- This allows developers to define a priority order in which banners should be shown. EG. If `BannerA` is a higher priority than `BannerB` we can guarantee `BannerA` is shown even if `BannerB`'s conditional checks complete first.
- It should therefore be straightforward to reorder the priority of banners. eg. add banners to a list, and the order of that list is the priority of the banners.

**What else has changed?**

- I've update `breaking-news.js` so it can be prioritised alongside the other banners. To do this I've had to update the module to export `{ show, canShow }` as `bannerPicker.js` expects any banner it  receives to match this API.
- `breaking-news.js` no longer needs to emit `mediator` events. eg. `mediator.emit('modules:onwards:breaking-news:ready', true)`. This was only emitted to notify the 5 banners above whether there was a breaking news alert - because breaking news was a higher priority. Now `bannerPicker.js` manages this prioritisation the 5 banner's listed above no longer need to listed for this `mediator` event.

### CODE REVIEW

**Comment by GHaberis** [comment]
@AWare @walaura @jranks123 

I've created a `bannerPicker.js` module as described above. Could you convert the modules which are yours from the list above so they export they required methods `canShow` and `show` and import them into `bannerPicker.js`? I'll convert any that are leftover.

Any comments/opinions are welcomed!

**Comment by walaura** [comment]
Amazing work! 👏

File: static/src/javascripts/projects/common/modules/ui/bannerPicker.js [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    canShow: () => Promise<boolean>,
+    show: () => void,
+};
+
+let banners: Array<Banner> = [breakingNews, optInEngagementAlert];
+
+let results: Array<'pending' | boolean> = new Array(banners.length).fill(
+    'pending',
+    0
+);
```
Comment (SiAdcock) [inline static/src/javascripts/projects/common/modules/ui/bannerPicker.js]: Could we initialise these variables in `init`? Then we could avoid the side effects of loading this module, and we wouldn't need to expose a function just for testing

File: static/src/javascripts/projects/common/modules/ui/bannerPicker.spec.js [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+        setTimeout(() => {
+            resolve(true);
+        }, 2500);
+    });
+const extraLongFalseCheck = () =>
+    new Promise(resolve => {
+        setTimeout(() => {
+            resolve(false);
+        }, 2500);
+    });
```
Comment (SiAdcock) [inline static/src/javascripts/projects/common/modules/ui/bannerPicker.spec.js]: I'd be hesitant to add tests that could take several seconds to execute (if that's what is happening here!). Is there a way to rewrite using Jest's [`advanceTimersByTime`](https://facebook.github.io/jest/docs/en/timer-mocks.html#advance-timers-by-time) method?

File: static/src/javascripts/projects/common/modules/ui/bannerPicker.js (hunk 1 above)
Comment (GHaberis) [reply static/src/javascripts/projects/common/modules/ui/bannerPicker.js]: `banners` and `results` sit outside the scope of `init` as `resetChecks` needs to be able to reset their values when this module is under test, `getSuccessfulBannerIndex` could go inside `init` though.

File: static/src/javascripts/projects/common/modules/ui/bannerPicker.spec.js (hunk 2 above)
Comment (GHaberis) [reply static/src/javascripts/projects/common/modules/ui/bannerPicker.spec.js]: A fair point, I'll investigate timer methods offered by Jest.

File: static/src/javascripts/projects/common/modules/ui/bannerPicker.js (hunk 1 above)
Comment (SiAdcock) [reply static/src/javascripts/projects/common/modules/ui/bannerPicker.js]: You could still define the variables in the top scope, but then use `init` to initialise them?

Comment (GHaberis) [reply static/src/javascripts/projects/common/modules/ui/bannerPicker.js]: OK I see what you meant, yes I can do this!

Comment (GHaberis) [reply static/src/javascripts/projects/common/modules/ui/bannerPicker.js]: On second thought, if I initialise them in `init` it'll override the values whenever I run `init` from the test module.

Comment (SiAdcock) [reply static/src/javascripts/projects/common/modules/ui/bannerPicker.js]: Good point. It sort of indicates that the banners shouldn't to be imported in this module. Perhaps they should be passed from the place where `bannerPicker.init` is called (`enhanced/common`). That would keep this module purer, make it more testable, and pass the banner-loading back up to the naturally more imperative `enhanced/common` module.

File: static/src/javascripts/bootstraps/enhanced/common.js [hunk 3]
```diff
@@ -283,25 +271,31 @@ const initialiseEmail = (): void => {
...
+        config.get('switches.idAdConsents', false)
+            ? firstPvConsentBanner
+            : cookiesBanner,
+        breakingNews,
+        smartAppBanner,
+        membershipEngagementBanner,
+        optInEngagementBanner,
+        signInEngagementBanner,
+        membershipBanner,
+    ];
```
Comment (GHaberis) [inline static/src/javascripts/bootstraps/enhanced/common.js]: Now all banners are converted in this branch we can decide on a priority order.

File: static/src/javascripts/bootstraps/enhanced/common.js [hunk 4]
```diff
@@ -283,25 +271,31 @@ const initialiseEmail = (): void => {
...
-        initCookiesBanner();
-    }
+const initialiseBanner = (): void => {
+    // ordered by priority
+    const bannerList = [
+        config.get('switches.idAdConsents', false)
+            ? firstPvConsentBanner
+            : cookiesBanner,
+        breakingNews,
+        smartAppBanner,
```
Comment (jranks123) [inline static/src/javascripts/bootstraps/enhanced/common.js]: As discussed in person, I would check whether we still want to be pointing new mobile web readers to the app

**Review by jranks123 — APPROVED** [review]

File: static/src/javascripts/projects/common/modules/onward/breaking-news.js [hunk 5]
```diff
@@ -154,64 +159,72 @@ const renderSpectre = ($breakingNews: bonzo): bonzo =>
...
+                    reportError(ex, {
+                        feature: 'breaking-news',
+                    });
+                    resolve(false);
                 });
-            });
-    }
-    return Promise.reject(new Error('cannot dismiss'));
+        } else {
+            resolve(false);
```
Comment (SiAdcock) [inline static/src/javascripts/projects/common/modules/onward/breaking-news.js]: Could return `Promise.resolve(false)` here, and within the `Promise` chain above return `true` or `false` as necessary. Then we don't need the `new Promise(...)` wrapper

File: static/src/javascripts/projects/common/modules/ui/bannerPicker.js [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+                    const trackingObj = {
+                        component: 'banner-picker',
+                        value: successfulBanner.id,
+                    };
+                    ophan.record(trackingObj);
+                }
+
+                if (!results.includes('pending')) {
+                    resolve();
```
Comment (SiAdcock) [inline static/src/javascripts/projects/common/modules/ui/bannerPicker.js]: Do we need to clear the `timeout` at this point? Not sure, just thinking memory leaks

Comment (GHaberis) [reply static/src/javascripts/projects/common/modules/ui/bannerPicker.js]: The `timeout` has already been cleared at this point. It's being cleared in the `then` block following `banner.canShow`.

File: static/src/javascripts/projects/common/modules/onward/breaking-news.spec.js [hunk 7]
```diff
@@ -14,340 +11,568 @@ jest.mock('lib/storage', () => ({
...
 
     afterEach(() => {
-        mediator.removeAllListeners();
-
-        $('.js-breaking-news-placeholder').remove();
+        localStorageStub.get.mockReset();
+        localStorageStub.isAvailable.mockReset();
+        fakeRelativeDates.isWithinSeconds.mockReset();
+        fakeFetchJson.mockReset();
+        fakeTemplate.mockReset();
```
[the remark this answers is not in the document]
Comment (GHaberis) [reply static/src/javascripts/projects/common/modules/onward/breaking-news.spec.js]: @SiAdcock I've made the suggested updates to `breaking-news.spec.js` we discussed last Friday.

**Review by SiAdcock — APPROVED** [review]

**Pull request merged**